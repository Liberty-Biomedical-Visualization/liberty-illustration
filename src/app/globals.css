@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
