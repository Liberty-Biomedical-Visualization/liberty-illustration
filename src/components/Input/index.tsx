import { type ChangeEvent, type HTMLInputTypeAttribute, useId } from "react";

import resolveClassNames from "@/lib/resolveClassNames";

export default function Input(props: Readonly<InputProps>) {
  const { disabled, label, name, onChange, type, validation, value } = props;

  const className = resolveClassNames(
    "flex flex-col relative pb-4",
    props.className,
  );

  const id = useId();
  const validationId = `${id}-validation`;

  const control =
    type === "textarea" ? (
      <textarea
        aria-describedby={validation ? validationId : undefined}
        className="border p-2"
        disabled={disabled}
        id={id}
        name={name}
        onChange={onChange}
        required
        value={value}
      />
    ) : (
      <input
        aria-describedby={validation ? validationId : undefined}
        className="border p-2"
        disabled={disabled}
        id={id}
        name={name}
        onChange={onChange}
        required
        type={type}
        value={value}
      />
    );

  return (
    <div className={className}>
      <label htmlFor={id}>{label}</label>
      {control}
      {validation && (
        <div
          className="absolute text-red-600 text-sm -bottom-1"
          id={validationId}
          role="alert"
        >
          {validation}
        </div>
      )}
    </div>
  );
}

export interface InputProps {
  className?: string;
  disabled?: boolean;
  label: string;
  name: string;
  onChange?: (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
  ) => void;
  type: HTMLInputTypeAttribute | "textarea";
  validation?: string | null;
  value?: string;
}
