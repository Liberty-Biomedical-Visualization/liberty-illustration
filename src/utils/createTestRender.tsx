import { render } from "@testing-library/react";
import type { JSX, JSXElementConstructor } from "react";

/**
 * Returns a function that allows rendering Component with partial props while
 * falling back to defaultProps for testing purposes.
 */
export default function createTestRender<C extends JSXElementConstructor<P>, P>(
  Component: JSXElementConstructor<P> & C,
  defaultProps: JSX.LibraryManagedAttributes<C, P>,
) {
  return function (props: Partial<P> = {}) {
    const resolvedProps: JSX.IntrinsicAttributes &
      JSX.LibraryManagedAttributes<C, P> = { ...defaultProps, ...props };
    return render(<Component {...resolvedProps} />);
  };
}
