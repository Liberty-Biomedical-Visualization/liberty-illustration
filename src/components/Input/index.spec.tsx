import { describe, expect, it, jest } from "@jest/globals";
import { screen } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";

import createTestRender from "@/lib/createTestRender";

import Input from ".";

describe("Input", () => {
  it("should display a form control", () => {
    renderInput({ type: "text" });
    const result = screen.getByRole("textbox");
    expect(result).toBeVisible();
  });

  it("should display a textarea form control when type is textarea", () => {
    const { container } = renderInput({ type: "textarea" });
    const result = container.querySelector("textarea");
    expect(result).toBeInTheDocument();
  });

  it("should display a form control with label as accessible name", () => {
    const label = "Text";
    renderInput({ label, type: "text" });
    const result = screen.getByRole("textbox", { name: label });
    expect(result).toBeVisible();
  });

  it("should call onChange when form control value changes", async () => {
    const onChange = jest.fn();
    const label = "Text";
    renderInput({ label, onChange, type: "text" });

    const user = userEvent.setup();
    const input = screen.getByLabelText(label);
    await user.type(input, "Foo");

    expect(onChange).toHaveBeenCalled();
  });

  it("should display a validation message when passed validation", async () => {
    const validation = "Bad value";
    renderInput({ validation });
    const validationMessage = screen.getByText(validation);
    expect(validationMessage).toBeInTheDocument();
  });

  it("should display a validation message for a textarea when passed validation", async () => {
    const validation = "Bad value";
    renderInput({ type: "textarea", validation });
    const validationMessage = screen.getByText(validation);
    expect(validationMessage).toBeInTheDocument();
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderInput({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderInput();
    expect(result.container).toMatchSnapshot();
  });
});

const renderInput = createTestRender(Input, {
  label: "Foo",
  name: "foo",
  type: "text",
});
