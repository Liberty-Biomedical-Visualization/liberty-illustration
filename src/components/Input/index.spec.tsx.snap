// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match the snapshot 1`] = `
<div>
  <div
    class="flex flex-col pb-4 relative"
  >
    <label
      for=":r7:"
    >
      Foo
    </label>
    <input
      class="border p-2"
      id=":r7:"
      name="foo"
      required=""
      type="text"
      value=""
    />
  </div>
</div>
`;
